### PR TITLE
Adding cgroup-parent option for docker build

### LINKED ISSUE #11822: Add `--cgroup-parent` option to `docker build`
The `--cgroup-parent` was recently added to `docker run` and `docker create`, would be great to have this available in `docker build` as well.

### PR DESCRIPTION
**Pull request by jbarbier**
Happy birthday Docker!
Thanks to @crosbymichael @nathanleclaire @jfrazelle

Closes #11822 

Signed-off-by: Julien Barbier write0@gmail.com

### CODE REVIEW

**Comment by duglin** [comment]
@jbarbier can you add some docs and testcases for this too?
docs are in docs/  and look in integration-cli for the tests

**Comment by jbarbier** [comment]
@duglin Added docs

**Comment by duglin** [comment]
Looking good so far.  Please squash your commits down to one.

**Comment by jbarbier** [comment]
@duglin Sorry, I don't know what it means :)

**Comment by duglin** [comment]
@jbarier there are two things:
1 - as you make additional changes, rather than doing a new commit, go ahead and reuse the previous one via: `git commit --amend`. That will keep you from having more than one commit.
2 - right now you have two commits, to squash (or merge) them into one do: `git rebase -i master`. When in the editor choose to squash the 2nd one by replacing the first word on that 2nd line with the word 'squash' or just 's' (replace 'pick').  It should them merge those two commits into one.

**Comment by vishh** [comment]
@jbarbier: Thanks for adding this feature. Overall LGTM. Can you add an integration test? Take a look at [this](https://github.com/docker/docker/blob/master/integration-cli/docker_cli_run_unix_test.go#L112) test as an example.

**Comment by mtesselH** [comment]
Added a test under build tests in test-integration-cli; It is a basic test only. PR on jbarbier ready to merge. 
Idea for future improvement would be run a container with the build image and do the same checks that the run test referenced by vishh does.

**Comment by nathanleclaire** [comment]
update: https://github.com/jbarbier/docker/pull/2/commits waiting for merge due to tests failing `gofmt`

File: docs/sources/reference/commandline/cli.md [hunk 1]
```diff
@@ -767,6 +768,9 @@ you refer to it on the command line.
 > children) for security reasons, and to ensure repeatable builds on remote
 > Docker hosts. This is also the reason why `ADD ../file` will not work.
 
+`docker build` has a `--cgroup-parent` option that causes the containers used
```
Comment (jamtur01) [inline docs/sources/reference/commandline/cli.md]: The ....

File: docs/sources/reference/commandline/cli.md [hunk 2]
```diff
@@ -767,6 +768,9 @@ you refer to it on the command line.
...
+in the build to be run with this option.
```
Comment (jamtur01) [inline docs/sources/reference/commandline/cli.md]: But what does it do? This merely says "to be run with this option"...

**Comment by vishh** [comment]
@jbarbier: Thanks for adding the test. Code LGTM.

**Comment by cpuguy83** [comment]
Needs go-fmt

**Comment by nathanleclaire** [comment]
@jbarbier just merge https://github.com/jbarbier/docker/pull/2 😉

**Comment by rhatdan** [comment]
I just wrote this exact patch.  Then discovered this pull request.  It looks like the patches need to be consolidated and then it can be merged. @jbarbier are you still working on it?

File: integration-cli/docker_cli_build_test.go [hunk 3]
```diff
@@ -5597,3 +5597,30 @@ func TestBuildEmptyStringVolume(t *testing.T) {
 
 	logDone("build - empty string volume")
 }
+
+func TestBuildContainerWithCgroupParent(t *testing.T) {
+	testRequires(t, NativeExecDriver)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_build_test.go]: Looks like this also needs `testRequires(t, SameHostDaemon)`

File: docs/sources/reference/commandline/cli.md (hunk 1 above)
Comment (nathanleclaire) [inline docs/sources/reference/commandline/cli.md]: @jamtur01 With this do you mean you want "The `docker build`"...?

Comment (jamtur01) [inline docs/sources/reference/commandline/cli.md]: Yes don't start a sentence with a command.

**Comment by thaJeztah** [comment]
@jbarbier DCO failed and it looks like this needs to be squashed?

**Comment by nathanleclaire** [comment]
I'm not sure what's up with the integration tests, it doesn't seem related to this PR (although I could be mistaken):

```
--- FAIL: TestContainerApiCommit (1.39s)
	docker_utils.go:624: "run 7e16ea2c13b36c9c9132045128060720fa1a99f1877d9b03623c608e889d1ac4 ls /test" failed with errors: ls: /test: No such file or directory
		, exit status 1
```

@jamtur01 @duglin @jfrazelle can you please review this patch?  It's been updated to improve the docs and fix the broken test.

**Comment by jamtur01** [comment]
Docs LGTM

**Comment by nathanleclaire** [comment]
Integration tests look good now, must have been an upstream issue.

**Comment by jfrazelle** [comment]
needs rebase but then I think we should be good for merge :)

**Comment by thaJeztah** [comment]
@jfrazelle think it still needs to be squashed; it's 7 commits right now? (Ok, they're from several contributors)

**Comment by nathanleclaire** [comment]
Rebased but it say we use the wrong test infrastructure:

```
integration-cli/docker_cli_build_test.go
```

What changed?

**Comment by jfrazelle** [comment]
We use gocheck now :( check out some of the others should be pretty easy
change

**Comment by nathanleclaire** [comment]
Yup that's what I've done.. keep finding little things I've missed but it'll get there

**Comment by nathanleclaire** [comment]
image

**Comment by jfrazelle** [comment]
LGTM

**Comment by moxiegirl** [comment]
LGTM

File: docs/sources/reference/commandline/cli.md [hunk 4]
```diff
@@ -847,6 +848,11 @@ you refer to it on the command line.
 > children) for security reasons, and to ensure repeatable builds on remote
 > Docker hosts. This is also the reason why `ADD ../file` will not work.
 
+When `docker build` is run with the `--cgroup-parent` option the containers used
+in the build will be run with the [corresponding `docker run`
+flag](/reference/run/#specifying-custom-cgroups).
```
Comment (thaJeztah) [inline docs/sources/reference/commandline/cli.md]: Not entirely happy with this sentence; I had some trouble understanding what was meant here, and almost interpreted *"will be run with the corresponding docker run flag"* as *"any docker run flag can be set here"*, (e.g. `--cgroup-parent=volumes-from=something`).

Perhaps something like; When docker build is run with the `--cgroup-parent`, the containers used in the build are run with the specified cgroup. For more information, see "Specifying custom cgroups"


**note** I just noticed that the `--cgroup-parent` flag for `docker run` [isn't documented](https://github.com/docker/docker/blob/8bd5a95e1e39541dfc5dc635c5c8e7604fe10028/docs/sources/reference/commandline/cli.md#run) in the command-line reference

File: docs/sources/reference/run.md [hunk 5]
```diff
@@ -465,6 +465,13 @@ Note:
 
 You would have to write policy defining a `svirt_apache_t` type.
 
+## Specifying custom cgroups
+
+Using the `--cgroup-parent` flag, you can pass a specific cgroup to run a
+container in. This allows you to create and manage cgroups on their own. You can
+define custom resources for those cgroups and put containers under a common
+parent group.
```
Comment (thaJeztah) [inline docs/sources/reference/run.md]: Do we need an actual example somewhere?

**Comment by SvenDowideit** [comment]
I too would like more information in the docs, but as we have 2 LGTM's, maybe @thaJeztah or someone can make a separate PR?

semi-LGTM?

**Comment by thaJeztah** [comment]
@SvenDowideit created a separate issue for improving the docs here https://github.com/docker/docker/issues/12849.

Does this PR still need to be squashed before merging (I know, I asked before)?

**Comment by nathanleclaire** [comment]
Nice job taking the initiative on that issue @thaJeztah .

> Does this PR still need to be squashed before merging (I know, I asked before)?

I can do this if absolutely necessary, but I am not sure how exactly it is supposed to work in a multi-author context with respect to DCOs etc.

**Comment by thaJeztah** [comment]
> Nice job taking the initiative on that issue @thaJeztah .

I only created it  for now, not sure when I have time, so others are more than welcome to start writing :)

> I can do this if absolutely necessary, but I am not sure how exactly it is supposed to work in a multi-author context with respect to DCOs etc.

Multiple sign-off messages on a commit should not be a problem.

@icecrime (being "Mr. Squash"); do you want the commits to be squashed, or is it fine as-is?

File: docs/sources/reference/commandline/cli.md (hunk 4 above)
Comment (nathanleclaire) [inline docs/sources/reference/commandline/cli.md]: This PR adds a little documentation to that run section which, while admittedly light on coverage, is better than nothing.

**Comment by thaJeztah** [comment]
ping @icecrime https://github.com/docker/docker/pull/11844#issuecomment-97205790, merge or squash first?

**Comment by nathanleclaire** [comment]
image

**Comment by thaJeztah** [comment]
\o/ thanks all!

**Comment by jbarbier** [comment]
ryu-shoryuken-xm

**Pull request merged**